frame/support/procedural/src/benchmark.rs

```diff
@@ -152,6 +173,41 @@ struct BenchmarkDef {
 impl BenchmarkDef {
 	/// Constructs a [`BenchmarkDef`] by traversing an existing [`ItemFn`] node.
 	pub fn from(item_fn: &ItemFn, extra: bool, skip_meta: bool) -> Result<BenchmarkDef> {
+		let invalid_return = |span| {
+			return Err(Error::new(
+				span,
+				"Invalid return type. Only `()` and `Result<(), BenchmarkError>` are allowed.",
+			))
+		};
+		if let ReturnType::Type(_, typ) = &item_fn.sig.output {
+			// parsing for non `()` return type (only allow `Result<(), BenchmarkError>`)
+			let Type::Path(TypePath { path, qself: _ }) = &**typ else { return invalid_return(typ.span()) };
+			let Some(segment) = path.segments.last() else { return invalid_return(path.span()) };
+			let _: keywords::Result = syn::parse(segment.ident.to_token_stream().into())?;
+			let PathArguments::AngleBracketed(args) = &segment.arguments else { return invalid_return(segment.span()) };
+			let Some(last_arg) = args.args.last() else { return invalid_return(args.args.span()) };
+			if args.args.len() > 2 {
+				return invalid_return(last_arg.span())
+			}
+			let Some(first_arg) = args.args.first() else { return invalid_return(args.args.span()) };
+			let GenericArgument::Type(Type::Tuple(TypeTuple { paren_token: _, elems })) = first_arg else { return invalid_return(first_arg.span()) };
+			if !elems.is_empty() {
+				return invalid_return(elems.span())
+			}
+			let _: keywords::BenchmarkError = syn::parse(last_arg.to_token_stream().into())?;
+		} else {
+			// return type is `()` which is permitted
+			// check for question mark ops as these aren't allowed here
+			let mut visitor = ExprTryVisitor::new();
+			visitor.visit_item_fn(item_fn);
+			if let Some(expr) = visitor.results.first() {
+				return Err(Error::new(
+					expr.question_token.span,
+					"The `?` operator can only be used in benchmark functions that specify `Result<(), BenchmarkError>`as their return type"
```

Comment:
Why is this needed? The rust compiler should already mention that the return type is invalid, or?

Reply:
Because in reality you could put any return type on the benchmark function definition and it would compile, even non-sensical ones, because the return type is discarded and replaced by the return type of the traits that get impl'ed. So this whole PR is just making the benchmark function definitions "feel" more like real functions by making the return type follow normal rust semantics to a certain extent